src/openrct2-ui/windows/EditorObjectSelection.cpp

```diff
@@ -1005,30 +1026,33 @@ static std::vector<Widget> _window_editor_object_selection_widgets = {
                 }
             }
 
-            const int32_t ride_tabs[] = {
-                SPR_G2_INFINITY,        SPR_TAB_RIDES_TRANSPORT_0, SPR_TAB_RIDES_GENTLE_0, SPR_TAB_RIDES_ROLLER_COASTERS_0,
-                SPR_TAB_RIDES_THRILL_0, SPR_TAB_RIDES_WATER_0,     SPR_TAB_RIDES_SHOP_0,   SPR_TAB_FINANCES_RESEARCH_0,
-            };
-            const int32_t ThrillRidesTabAnimationSequence[] = {
+            constexpr int32_t ThrillRidesTabAnimationSequence[] = {
                 5, 6, 5, 4, 3, 2, 1, 0, 0, 0, 0, 0, 0, 0, 1, 2, 3, 4, 0, 0, 0,
             };
 
-            // Draw ride tabs
-            if (GetSelectedObjectType() == ObjectType::Ride)
+            // Draw sub-tab images, if applicable
+            auto& currentPage = ObjectSelectionPages[selected_tab];
+            if (currentPage.Flags & ObjectPageFlags::hasSubTabs)
             {
-                for (int32_t i = 0; i < 7; i++)
+                for (auto i = 0u; i < currentPage.subTabs.size(); i++)
                 {
                     const auto& widget = widgets[WIDX_FILTER_RIDE_TAB_ALL + i];
                     if (widget.type == WindowWidgetType::Empty)
                         continue;
 
-                    int32_t spriteIndex = ride_tabs[i];
+                    auto& subTabDef = currentPage.subTabs[i];
+                    int32_t spriteIndex = subTabDef.baseImage;
                     int32_t frame = 0;
-                    if (i != 0 && IsWidgetPressed(WIDX_FILTER_RIDE_TAB_ALL + i))
+                    if (subTabDef.animationLength > 1 && _selectedSubTab == i)
                     {
-                        frame = frame_no / window_editor_object_selection_animation_divisor[i - 1];
+                        frame = frame_no / subTabDef.animationDivisor;
                     }
-                    spriteIndex += (i == 4 ? ThrillRidesTabAnimationSequence[frame] : frame);
+
+                    // TODO: generalise this?
+                    if (currentPage.Caption == STR_OBJECT_SELECTION_RIDE_VEHICLES_ATTRACTIONS && i == 4)
+                        spriteIndex += ThrillRidesTabAnimationSequence[frame];
```

Comment:
Not sure what to do with this yet. It's rare enough to not bloat the `ObjectSelectionPages` struct with it, I think. Leaving as-is for now.